In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

Load core datasets

In [44]:
base_path = "../data/raw/california/"

incidents = pd.read_csv(base_path + "NIBRS_incident.csv")
victims = pd.read_csv(base_path + "NIBRS_VICTIM.csv")
offenders = pd.read_csv(base_path + "NIBRS_OFFENDER.csv")
offenses = pd.read_csv(base_path + "NIBRS_OFFENSE.csv")
arrestees = pd.read_csv(base_path + "NIBRS_ARRESTEE.csv")
agencies = pd.read_csv(base_path + "agencies.csv", encoding="cp1252")

Load lookup tables

In [45]:
weapon_type = pd.read_csv(base_path + "NIBRS_WEAPON_TYPE.csv")
race = pd.read_csv(base_path + "REF_RACE.csv")
ethnicity = pd.read_csv(base_path + "NIBRS_ETHNICITY.csv")
age = pd.read_csv(base_path + "NIBRS_AGE.csv")
location_type = pd.read_csv(base_path + "NIBRS_LOCATION_TYPE.csv")
relationship = pd.read_csv(base_path + "NIBRS_RELATIONSHIP.csv")
state = pd.read_csv(base_path + "REF_STATE.csv")
offense_type = pd.read_csv(base_path + "NIBRS_OFFENSE_TYPE.csv")


Secondary tables

In [46]:
victim_offender_rel = pd.read_csv(base_path + "NIBRS_VICTIM_OFFENDER_REL.csv")
victim_injury = pd.read_csv(base_path + "NIBRS_VICTIM_INJURY.csv")
victim_circumstances = pd.read_csv(base_path + "NIBRS_VICTIM_CIRCUMSTANCES.csv")
weapon = pd.read_csv(base_path + "NIBRS_WEAPON.csv")
nibrs_property = pd.read_csv(base_path + "NIBRS_PROPERTY.csv")

Verification

In [47]:
print("incidents:", incidents.shape)
print("victims:", victims.shape)
print("offenders:", offenders.shape)
print("arrestees:", arrestees.shape)
print("agencies:", agencies.shape)
print("offenses:", offenses.shape)

incidents: (1188704, 15)
victims: (1362826, 16)
offenders: (1351022, 11)
arrestees: (370032, 18)
agencies: (655, 59)
offenses: (1369571, 8)


## Data Inspection

### Incidents Dataset

In [52]:
print("Head: ")
incidents.head()

Head: 


,data_year,agency_id,incident_id,nibrs_month_id,cargo_theft_flag,submission_date,incident_date,report_date_flag,incident_hour,cleared_except_id,cleared_except_date,incident_status,data_home,orig_format,did
0,2024,937,190226279,44856163,f,2024-05-02 20:04:52.306,2024-03-05,f,14.0,6,NaN,ACCEPTED,NaN,X,246196887
1,2024,937,185618101,44732720,f,2024-02-08 04:53:04.531,2024-01-01,f,5.0,6,NaN,ACCEPTED,NaN,X,231242773
2,2024,937,187487274,44732720,f,2024-03-11 01:20:17.253,2024-01-01,f,5.0,6,NaN,WARNINGS,NaN,X,237181432
3,2024,937,185618105,44732720,f,2024-02-08 04:53:05.521,2024-01-01,f,8.0,6,NaN,ACCEPTED,NaN,X,231242785
4,2024,937,185618107,44732720,f,2024-02-08 04:53:06.053,2024-01-01,f,10.0,6,NaN,ACCEPTED,NaN,X,231242791


In [53]:
print("Shape: ")
incidents.shape

Shape: 


(1188704, 15)

In [54]:
print("Info: ")
incidents.info()

Info: 
<class 'pandas.DataFrame'>
RangeIndex: 1188704 entries, 0 to 1188703
Data columns (total 15 columns):
 #   Column               Non-Null Count    Dtype  
---  ------               --------------    -----  
 0   data_year            1188704 non-null  int64  
 1   agency_id            1188704 non-null  int64  
 2   incident_id          1188704 non-null  int64  
 3   nibrs_month_id       1188704 non-null  int64  
 4   cargo_theft_flag     1188704 non-null  str    
 5   submission_date      1188704 non-null  str    
 6   incident_date        1188704 non-null  str    
 7   report_date_flag     1188704 non-null  str    
 8   incident_hour        1161668 non-null  float64
 9   cleared_except_id    1188704 non-null  int64  
 10  cleared_except_date  9972 non-null     str    
 11  incident_status      1188704 non-null  str    
 12  data_home            0 non-null        float64
 13  orig_format          1188704 non-null  str    
 14  did                  1188704 non-null  int64  
dtypes:

In [55]:
print("Describe: ")
incidents.describe()

Describe: 


,data_year,agency_id,incident_id,nibrs_month_id,incident_hour,cleared_except_id,data_home,did
count,1188704.0,1.188704e+06,1.188704e+06,1.188704e+06,1.161668e+06,1.188704e+06,0.0,1.188704e+06
mean,2024.0,2.005169e+03,1.972396e+08,4.493933e+07,1.270263e+01,5.972773e+00,NaN,2.688154e+08
std,0.0,4.468076e+03,5.924535e+06,1.028431e+05,6.688144e+00,3.105171e-01,NaN,1.903355e+07
min,2024.0,9.370000e+02,1.831516e+08,4.468028e+07,0.000000e+00,1.000000e+00,NaN,2.235317e+08
25%,2024.0,1.205000e+03,1.924747e+08,4.489581e+07,8.000000e+00,6.000000e+00,NaN,2.537818e+08
50%,2024.0,1.387000e+03,1.975485e+08,4.495744e+07,1.400000e+01,6.000000e+00,NaN,2.699755e+08
75%,2024.0,1.582000e+03,2.020021e+08,4.501594e+07,1.800000e+01,6.000000e+00,NaN,2.841711e+08
max,2024.0,4.288200e+04,2.084446e+08,4.516236e+07,2.300000e+01,6.000000e+00,NaN,3.043163e+08


In [56]:
print("Sum of null : ")
incidents.isnull().sum()

Sum of null : 


data_year                    0
agency_id                    0
incident_id                  0
nibrs_month_id               0
cargo_theft_flag             0
submission_date              0
incident_date                0
report_date_flag             0
incident_hour            27036
cleared_except_id            0
cleared_except_date    1178732
incident_status              0
data_home              1188704
orig_format                  0
did                          0
dtype: int64

#### Incidents Observations

- Grain of dataset: Each row represents one crime incident reported in California (2024). No aggregation is done at row level.

- Key identifier: incident_id (used to uniquely identify each incident and link with other tables)

- Time-related columns:
  - submission_date → when the record was submitted (string, needs datetime conversion)
  - incident_date → when the crime happened (string, needs datetime conversion)
  - incident_hour → hour of incident (partially missing ~27,036 values)
  - nibrs_month_id → encoded month identifier
  - cleared_except_date → date of case clearance (very sparse)

- Missing values:
  - incident_hour → missing for some records (unknown time)
  - cleared_except_date → mostly missing (cases not cleared or not updated)
  - data_home → completely empty (no usable information)

- Column types issue:
  - Dates are stored as strings (not datetime yet)
  - Most IDs are integers (used for linking, not calculation)

- System/unused fields:
  - data_home → empty column (no data present)
  - orig_format → data source format indicator (not used for analysis)
  - did → internal system identifier (not needed for analysis)